### PR TITLE
Implement low-level MMV API and mmvdump

### PR DESCRIPTION
**Pull request by saurvs**
This will also be useful for implementing the `mmvdump` command and for testing MMV writing.

- [x] Print string repr for unit
- [x] Print string repr for mmv flags
- [x] Check for invalid indom id
- [x] Check for invalid item id

### CODE REVIEW

**Comment by suyash** [comment]
I actually considered this for speed, but didn't see a utility for this, since you cannot make things like indom offsets public fields. Why not implement mmvdump directly as a binary?

**Comment by suyash** [comment]
Ideally we should be able to do `cargo install mmvdump` to get the rust mmvdump binary

**Comment by saurvs** [comment]
@suyash This would also be a better way to test the bytes of the MMV after we export it.

**Comment by saurvs** [comment]
If we put the mmvdump implementation in `src/bin`, we can do `cargo install hornet` and then run `mmvdump <file>`.

**Comment by suyash** [comment]
Oh, sorry that slipped my mind, so you'll add client tests separately?

**Comment by suyash** [comment]
Let's hold off on `cargo install` for now, if you'll add client tests separately then this looks good to me

**Comment by suyash** [comment]
Also, test this against fixed cases, some are at https://github.com/performancecopilot/speed/tree/master/mmvdump/testdata

**Comment by saurvs** [comment]
Yes, once the indom branch is merged I'll update the tests in client.

**Comment by saurvs** [comment]
(TODOs moved to top comment)

**Comment by suyash** [comment]
@saurvs please document all public types, methods and fields, especially uses of `ffi::CStr`

File: tests/mmvdump.rs
Comment (suyash) [inline tests/mmvdump.rs]: don't do this, add a function to the library to export the string dump. Ideally `print_indoms` would be `write_indoms` taking `io::BufWriter` as a parameter

**Comment by saurvs** [comment]
Pinging @suyash @ryandoyle @cuviper for review.

**Comment by suyash** [comment]
@saurvs  add the todos to the PR's opening comment

**Comment by saurvs** [comment]
@suyash Also I don't think we need to document use of `CStr` since it's not a public parameter anywhere, and is only used internally to read null-terminated strings from the MMV.

**Comment by suyash** [comment]
@saurvs I insist, any special uses of `&str` vs `String` vs `CStr` should be added as a comment for anyone reading the code to understand, now or in the future

**Comment by suyash** [comment]
Just add a single line comment in front of `use ffi::CStr` as `// used to read null terminated strings`

File: src/client/metric.rs
Comment (ryandoyle) [inline src/client/metric.rs:105]: Can we have there numbers as constants? Ideally named the same as the `#define`s in `src/include/pcp/pmapi.h`?

See `src/include/pcp/pmapi.h:83`

File: src/client/metric.rs
Comment (ryandoyle) [inline src/client/metric.rs:151]: Same as previous comment, see `src/include/pcp/pmapi.h:91`

File: src/client/metric.rs
Comment (ryandoyle) [inline src/client/metric.rs:374]: As before, see `src/include/pcp/pmapi.h:129`

**Review by ryandoyle — CHANGES_REQUESTED** [review]
Looks good, just a couple of small things

File: tests/mmvfmt.rs
Comment (suyash) [inline tests/mmvfmt.rs]: This is not needed anymore

File: src/client/metric.rs
Comment (suyash) [reply src/client/metric.rs:105]: @ryandoyle  these are only convenience functions to create Units from their PMAPI representations (unsigned integers, as enums values map to them in C), however @saurvs has implemented the API more in tao of rust than C, and enums are first class values. This function is just a convenience that returns an enum value from an integer. He's also impl `fmt::Display` so you can directly print a `Space::Kilobyte` as `KiB` for debugging

File: src/mmv/mod.rs
Comment (suyash) [inline src/mmv/mod.rs:456]: This part is generally most familiar to people familiar with PMAPI, so please also add corresponding references to the `pmapi.h` file with line numbers

Comment (suyash) [reply src/mmv/mod.rs:456]: Please do this for all (possible) types in MMV module

**Review by suyash — APPROVED** [review]

File: src/mmv/mod.rs
Comment (suyash) [reply src/mmv/mod.rs:456]: Just add the GitHub url with line numbers for convenience

**Comment by suyash** [comment]
@saurvs for mmvdump add a test case that prints an inverted unit for example Hz from your first example, it would print it as `/s`. Please try to do this so I can merge this so we can get reviews in early. Good Work!

File: src/client/metric.rs
Comment (saurvs) [reply src/client/metric.rs:105]: @ryandoyle It would be better to force the use of enums over constants, even internally, since Rust mandates we handle all of its variants, and like @suyash said it also gives us the proper string representation.

Comment (ryandoyle) [reply src/client/metric.rs:105]: Yeah fair enough. Come to think of it, it would just be extra indirection having `0` defined as a constant when we know its a `Space::Byte` already.

File: src/client/metric.rs
Comment (suyash) [inline src/client/metric.rs]: remove this space, this is causing the output in test case 5 to be `MiB  / sec` (extra space after `MiB`)

File: src/client/metric.rs
Comment (suyash) [inline src/client/metric.rs:356]: add that space here, before `(`

**Comment by suyash** [comment]
@ryandoyle you need to approve for me to be able to merge this

File: src/client/metric.rs
Comment (saurvs) [reply src/client/metric.rs]: That would print adjacent units without separating spaces. I've instead removed the space before `/`. That would print `unit =/ ` in some cases, but it's how the C version of mmvdump also prints.

**Comment by suyash** [comment]
This is good to go from my end, @ryandoyle an unapproved review blocks the PR, please approve/add further comments

**Review by ryandoyle — APPROVED** [review]

**Comment by ryandoyle** [comment]
Hey @suyash, I've approved the review. Feel free to merge now.

Cheers

**Pull request merged**
